### DISCUSSION TITLE
Default value of `False` cast to `'False'` and is thus True

### DISCUSSION THREAD

**Discussion by jaraco** [fastapi/typer#942] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorials in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer


def main(do_bad_thing=False):
    if do_bad_thing:
        raise NotImplementedError


__name__ == '__main__' and typer.run(main)
```

### Description

When running a script without type annotations, but a default boolean value, I sort-of expected the type to be inferred from the default. That's how [autocommand](https://github.com/Lucretiel/autocommand/) works. If the default value is a boolean and no type annotation is given, the type is inferred to be `bool`. But in Typer, the type presumably defaults to `str`, so when running the above script with no parameters, `do_bad_thing` is set to `'False'`, which is `True`, and a `NotImplementedError` is raised.

Obviously, this unexpected behavior can be fixed by always supplying type annotations, but since the issue happens silently, it's a bit of a foot gun.

Better would be to either use the default value to infer that the type is a bool or warn or raise an Exception if the default value is a bool but no type annotation was given. It's unlikely if the user is supplying a default value of `True` or `False` that they want that cast to a string.

I note also that this behavior differs from how `arg = None` is handled, which apparently does not get cast to a string; it seems to be handled as `str | None` based on the default of `None`.

This issue came up in jaraco.develop, where after migrating to Typer in jaraco/jaraco.develop#21, I ran the update-projects script, but it failed when [`dry_run` was set to `'False'` (`True`)](https://github.com/jaraco/jaraco.develop/blob/08834e3074a35dda8fc313afc62b8b69dfa6677d/jaraco/develop/update-projects.py#L51).

### Operating System

macOS

### Operating System Details

n/a

### Typer Version

0.12.4

### Python Version

3.13.0rc1

**Comment by YuriiMotov** [comment, marked as answer]
You should specify the type:
```
def main(do_bad_thing: bool = False):
    ...
```

Then it works as expected
